// @fleek-platform/cli/dist/commands/functions/plugins/asyncLocalStoragePolyfill.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncLocalStoragePolyfill = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const asyncLocalStoragePolyfill = () => {
    return {
        name: 'replace-async-local-storage',
        setup(build) {
            build.onResolve({ filter: /async_hooks/ }, (args) => {
                if (args.path === 'async_hooks' || args.path === 'node:async_hooks') {
                    return {
                        path: node_path_1.default.resolve(__dirname, 'polyfills', 'async_hooks.js'),
                        namespace: 'replace-als',
                    };
                }
            });
            build.onLoad({ filter: /.*/, namespace: 'replace-als' }, async (args) => {
                const contents = await node_fs_1.promises.readFile(args.path, 'utf8');
                return {
                    contents,
                    loader: 'js',
                };
            });
        },
    };
};
exports.asyncLocalStoragePolyfill = asyncLocalStoragePolyfill;
