// @fleek-platform/cli/dist/commands/functions/plugins/unsupportedModuleStub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.moduleChecker = void 0;
const runtimeModules_1 = require("../runtimeModules");
const unsupportedModules = runtimeModules_1.unsupportedRuntimeModules.map((it) => `node:${it}`);
const moduleChecker = (args) => {
    const { unsupportedModulesUsed } = args;
    return {
        name: 'moduleChecker',
        setup: (build) => {
            build.onLoad({ filter: /.*/, namespace: 'unsupported' }, (args) => {
                console.log('unsupported', args);
                return {
                    contents: `
            throw new Error('Unsupported module: ${args.path}');
          `,
                    loader: 'js',
                };
            });
            build.onResolve({ filter: /.*/ }, ({ path }) => {
                if (unsupportedModules.includes(path) ||
                    unsupportedModules.includes(`node:${path}`)) {
                    unsupportedModulesUsed.add(path);
                    return {
                        path,
                        namespace: 'unsupported',
                    };
                }
                return null;
            });
        },
    };
};
exports.moduleChecker = moduleChecker;
